Implement Faro design

PeeterPaal opened the pull request:
## Description

Implements new, ecommerce oriented Faro design. This design uses a modular block system, that allows user to add/remove/change blocks and move them up or down in order, plus allows adding background colour or image to each block separately. Compared to other Voog designs, this design <ins>does not support</ins> Blog features or "old style" store with Product list and Product layouts. Product list layout is instead intended for use with Product list content area.

Current PR should accomplish fully-featured design that is ready for presenting to clients.

## Testing

On **Front page, Categories and Common page layouts**, check that switching between layouts works as expected and changes are saved correctly. Switching can be done by choosing a desired layout from dropdown visible in editmode.

<img width="266" alt="Screenshot 2023-11-09 at 16 35 04" src="https://github.com/Voog/design-faro/assets/44435196/3c40236c-10d9-452e-a822-461cad08e811">

On **Categories and Common page** layouts, verify that block system works as expected. Check that
- Adding blocks works. Adding a block always creates a block with default layout defined in `template-settings-json` that can then be modified. Block can be added below all other blocks by clicking on the grey `+` icon. 

Picture of add button, which is placed right before footer and below all other blocks.
<img width="1894" alt="Screenshot 2023-11-09 at 16 50 30" src="https://github.com/Voog/design-faro/assets/44435196/81019952-d2f0-4012-b73b-af3bf1625cf1">

- Removing a block works. Removing a block should not refresh the page, but a request should be made and removed from DOM with JS afterwards instead. Block can be removed by hovering on a block and clicking on trashcan icon in top-right.
- Changing block positions works. Changing positions should not refresh the page, but a request should be made and then switch should be done in DOM with JS afterwards. Moving can be done by hovering over a block and clicking on up/down button.

Picture of editmode design

<img width="1893" alt="Screenshot 2023-11-09 at 16 44 22" src="https://github.com/Voog/design-faro/assets/44435196/e1d62313-60e9-4ef0-9b89-5a24ebc57083">

Check that **Header** works. Expected behaviour in main menu is to hide all overflowing menu items and place them inside a dropdown menu. This should also work on window resize. Also verify mobile design.

Verify that **Design editor** works and settings change what they say that they change.

Verify that **Auto-generated Product page** works as expected, with buy buttons, sale prices, product images etc. present.

Verify that **BGPickers** work whereever they can be found. At least one BGPicker should be present on every page, with every block containing its own BGPicker. Blocks on Categories page have separate BGPickers for both columns.

Verify that **Search** is not displayed when it's turned off in site settings.

Verify that **Cart** button is not displayed when store is turned off for site.

### Files that can be omitted when doing a code review:
All `template-cs-*.tpl` files, as these are auto-generated from `sources/components/custom-styles/template-cs-*.scss` files.

`javascripts` and `stylesheets` folder contents, as these are auto-generated from `sources/javascripts` and `sources/stylesheets`.

`sources/javascripts/libs/colorextract.js`, as this is copied over from previous Voog designs.

PeeterPaal committed:
Add base files

PeeterPaal committed:
Add initial bgpicker to front page

PeeterPaal committed:
Show bgpicker button only in editmode

PeeterPaal committed:
Implement initial main menu

Implements main menu, that is sticky and has dropdown menus. Header
and dropdown menus use the same background color as page. To achieve
this, bgpicker will write RGB value to variable, that is then used in
CSS.

PeeterPaal committed:
Use default value for background RGB value

PeeterPaal committed:
Add split front page layout type

PeeterPaal committed:
Add button styles & design system colours

PeeterPaal committed:
Add base layouts

PeeterPaal committed:
Add initial categories page layout

PeeterPaal committed:
Implement front page layouts

Implements 3 different front page layouts - split, column and
column-split. Layout is applied by assigning according classname.

PeeterPaal committed:
Implement Product list layout

Implements Product list layout that includes default styles for
product list widget. Includes breadcrumb menu component & bgpicker.

PeeterPaal committed:
Fix header content area empty state

PeeterPaal committed:
Add initial auto-gen product layout

PeeterPaal committed:
Improve auto-gen product pages & buybuttons

- Add bgpicker to product page
- Add footnote area to product page
- Improve buy button placement visuals

PeeterPaal committed:
Implement dropdown menu for overflowing items

Adds dropdown menu that contains menuitems that do not fit the top bar.
Overflowing items are added to menu using JS and this runs on page load
and window resize.

PeeterPaal committed:
Add initial language menu

PeeterPaal committed:
Add max width to most layouts

PeeterPaal committed:
Add initial Common Page layout

Common page layout will be composed of different blocks, each block has
a slightly different layout. Currently 4 block layouts are supported.

PeeterPaal committed:
Enlarge top menu element min width

PeeterPaal committed:
Render block only if it has content

PeeterPaal committed:
Implement cart button and added-to-cart modal

PeeterPaal committed:
Add bgpickers to blocks

Add separate bgpicker to each block added to page. Blocks are now
read from and saved to page data. This makes it possible to add
additional blocks and/or remove or modify them.

PeeterPaal committed:
Save categories sections to page data

PeeterPaal committed:
Add bgpickers to category sections

PeeterPaal committed:
Only calculate cart items count if cart is present

PeeterPaal committed:
Implement footer

PeeterPaal committed:
Improve front page layout

- Make image area a content area
- Simplify contents layout
- Make hero area "main" content area

PeeterPaal committed:
Add mobile styles for front page layout

PeeterPaal committed:
Add mobile styles for auto-gen product page

PeeterPaal committed:
Add mobile styles for common page blocks

PeeterPaal committed:
Add mobile styles for footer

PeeterPaal committed:
Add mobile styles for categories page

PeeterPaal committed:
Use minified editmode JS file

PeeterPaal committed:
Add initial mobile menu

PeeterPaal committed:
Fix displaying submenus on desktop

PeeterPaal committed:
Improve menu handling

Make sure that at least one menu element is always present in top menu

PeeterPaal committed:
Center header content area in mobile view

PeeterPaal committed:
Make sure lang menu does not end up in main menu

PeeterPaal committed:
Refactor blocks logic

Instead of storing key-value pairs, store each block data in an array,
which includes a key field. Update bg picker scripts accordingly.
Use the same structure for both common page & category page blocks.

PeeterPaal committed:
Refactor blocks' rendering

Both common page and categories page now use the same block component
for rendering blocks.

Content area count is defined in block settings and is no longer static.

Block layouts are now names instead of numbered.

PeeterPaal committed:
Fix setting backgrounds for blocks

PeeterPaal committed:
Enable changing the order of blocks

Make changing the order of blocks available via UI by up/down buttons
that change the position of blocks and commit changes to page data.

PeeterPaal committed:
Allow adding additional blocks in editmode

Enable adding more blocks in editmode via separate buttons, which are
rendered based on allowed blocks for current layout. Adding new block
will try to resolve collisions for block keys by adding an additional
suffix to it.

PeeterPaal committed:
Allow deleting blocks in editmode

PeeterPaal committed:
Fix categories block height + dropdown positioning

PeeterPaal committed:
Fix front page width

PeeterPaal committed:
Improve language & cart btn position on mob. menu

PeeterPaal committed:
Add initial mobile submenu navigation

PeeterPaal committed:
Fix categories blocks in mobile view

PeeterPaal committed:
Add cart popover mobile styles

PeeterPaal committed:
Improve mobile menu styles

Use less JS and more CSS to position mobile menu elements

PeeterPaal committed:
Fix categories sections height in mobile

PeeterPaal committed:
Improve mobile menu handling

- Close submenus on menu close
- Remove back btn on menu close

PeeterPaal committed:
Add search modal

PeeterPaal committed:
Add initial mobile search

PeeterPaal committed:
Improve search modal & window resize behaviour

Only run resize events on window width change, in order to prevent
these functions running on mobile keyboard opening

PeeterPaal committed:
Use flexbox instead of grid for top menu

PeeterPaal committed:
Improve desktop submenu background

PeeterPaal committed:
Add the option to change block layout

PeeterPaal committed:
Simplify block action binding

PeeterPaal committed:
Fix front page width

PeeterPaal committed:
Improve mobile search behaviour

PeeterPaal committed:
Fix deleting category blocks

As content_class was passed to listener, passing multiple classnames
did not correcly initialize delete button listeners

PeeterPaal committed:
Add clear button to search form

PeeterPaal committed:
Add support for design editor

Configure taskfile to support building design editor components & add
initial custom styles

PeeterPaal committed:
Fix setting bg in front page layout

PeeterPaal committed:
Add initial darkmode colors

PeeterPaal committed:
Add base font files

PeeterPaal committed:
Add heading & content design editor styles

PeeterPaal committed:
Add button design editor styles

PeeterPaal committed:
Add header design editor styles

PeeterPaal committed:
Improve form styles

PeeterPaal committed:
Support darkmode in product widget content area

PeeterPaal committed:
Fix setting bg in product list layout

Initialize blockData variable if its not set in layout beforehand

PeeterPaal committed:
Fix sass module name

PeeterPaal committed:
Reduce form field sizes in mobile view

PeeterPaal committed:
Make current language not clickable

PeeterPaal committed:
Add main menu active design editor styles

PeeterPaal committed:
Add footer design editor styles

PeeterPaal committed:
Add darkmode styles for mobile search & menus

PeeterPaal committed:
Add product page design editor styles

PeeterPaal committed:
Improve buybtn styles

PeeterPaal committed:
Initialize blockdata var in product list layout

PeeterPaal committed:
Make applying dark bg more general

PeeterPaal committed:
Add form design editor styles

PeeterPaal committed:
Initialize blockData in auto-gen pr. layout

PeeterPaal committed:
Remove some featured design editor settings

PeeterPaal committed:
Improve mobile language menu

Position menu items under eachother & display arrow

PeeterPaal committed:
Remove form submit btn from form styles

PeeterPaal committed:
Always show subpage add btn in editmode

PeeterPaal committed:
Improve bgpicker button placement

PeeterPaal committed:
Remove row template from split front page layout

PeeterPaal committed:
Fix category blocks in mobile view

PeeterPaal committed:
Add editmode design for block actions

PeeterPaal committed:
Use single wrapper for categories blocks

PeeterPaal committed:
Add block add button editmode design

PeeterPaal committed:
Fix hover events on category blocks

PeeterPaal committed:
Remove icons as components & use them in CSS

PeeterPaal committed:
Fix block reordering and deletion

PeeterPaal committed:
Show block move btns only if applicable

If block is first, then it can only be moved down, and if last, then
it can only be moved upwards.

PeeterPaal committed:
Only show cart btn if store is active

PeeterPaal committed:
Add `Bodoni Moda` as default font

PeeterPaal committed:
Make content font adjustments more general

PeeterPaal committed:
Format JS files w/ Prettier

PeeterPaal committed:
Fix cart btn placement on mobile

PeeterPaal committed:
Add `sd_product` tag to auto-gen product layout

t-p-developer commented on components/menu-lang.tpl:
```diff
@@ -0,0 +1,13 @@
+  {% if editmode -%}
+    <span class="edit-btn">
+      {% languageadd %}
+    </span>
```
nit! `span` has a class `edit-btn`, why not use `button` then?

PeeterPaal replied:
This class was renamed afterwards, I believe this was from an earlier commit. Now this is wrapped in a `menu-child` class.
https://github.com/Voog/design-faro/blob/dec318fd3d81a95df237f0af6639957f5532a963/components/menu-lang.tpl#L30

`{% language %}` tag creates a styled link, so it is unnecessary to wrap inside a button.

t-p-developer commented on sources/stylesheets/_content-formatted.scss:
```diff
@@ -0,0 +1,132 @@
+  .edy-image-container-with-title:after {
+    display: block;
+    padding: 4px;
+    font-size: 12px;
```
nit! I personally prefer to use `rem`-s instead of `px`-s.

t-p-developer commented on components/header.tpl:
```diff
@@ -1,11 +1,13 @@
-<div class="header">
-  {% include "menu-main" %}
+<div class="header-wrapper">
```
nit! Maybe instead of using `header-wrapper` class we should use `header` tag.

t-p-developer commented on sources/stylesheets/_general.scss:
```diff
@@ -1,7 +1,13 @@
+  margin: 0;
+  padding: 0;
+}
+
```
nit! maybe we can add `normalize-css` into the mix, similar to https://github.com/necolas/normalize.css/blob/master/README.md?

t-p-developer commented on javascripts/editmode.js:
```diff
@@ -77,12 +78,15 @@
       }
 
+      if (data.colorData) {
+        colorRgb = data.colorData.r + ',' + data.colorData.g + ',' + data.colorData.b;
```
nit! If possible, use `template strings`.

PeeterPaal replied:
At first I tried to avoid using ES6 features to make this template as compatible as possible, but since other parts of the site would probably break on older browsers regardless, we allowed ES6 aswell. Good point, template literals would be good here.

PeeterPaal committed:
Inlcude image data in bgpicker dataset

t-p-developer commented on javascripts/application.js:
```diff
@@ -578,8 +598,50 @@
     };
 
+    const handleMenuContent = () => {
+      console.log('running');
```
nit! Is this here by accident?

PeeterPaal replied:
Indeed :)

t-p-developer commented on components/footer.tpl:
```diff
@@ -0,0 +1,16 @@
+<div class="footer">
```
nit! Maybe use the `footer` tag instead?

t-p-developer commented on sources/stylesheets/_variables.scss:
```diff
@@ -32,3 +32,5 @@ $z-index-dropdown: 11015;
 
 $max-page-width: 1440px;
+
+$mobile-width: 900px;
```
nit! Based on the size, it seems to be more `$table-width`.

t-p-developer commented on sources/stylesheets/components/_search.scss:
```diff
@@ -0,0 +1,95 @@
+  }
+
+  .search-inner {
+    max-width: 992px;
```
nit! I believe we have a variable for this `max-width`?

PeeterPaal replied:
Perhaps in some other design, but I decided against bringing it in here, as this max width is not widely used in this design.

t-p-developer reviewed (changes requested):
This would be the first round of feedbacks. In overall nicely done, especially component folder.

t-p-developer commented:
The `delete` and `arrow` buttons are far apart on the larger screen. Is it possible to bring them closer or add them in a `1440px` wrapper?

t-p-developer commented:
If the user has a `column`, then some content will be missing. It will not happen with `dominant right` or `dominant left`, is this expected?

PeeterPaal commented:
> The delete and arrow buttons are far apart on the larger screen

This should be discussed with designer, but I believe bringing these buttons closer could interfere with content areas, making some parts of them inaccessible. I already ran into some problems while implementing these.

> If the user has a column, then some content will be missing. It will not happen with dominant right or dominant left, is this expected?

Currently this is expected behaviour, as `column` type layout has only a singular content area, while other layouts have two content areas.

PeeterPaal committed:
Use `header` tag for header & improve classnames

PeeterPaal committed:
Use `footer` tag for footer

PeeterPaal committed:
Improve screen width variable names

PeeterPaal committed:
Use template literals

PeeterPaal committed:
Use darkmode color for front page layout select

piiamatsi reviewed (commented):
- If an image is added to the text area beside main menu, it doesn't display the dots by which the size of the image could be adjusted.

- When refreshing the page or switching to another one, the menu items hidden under the ellipsis are visible for a moment.

- A page with Product list or Common page layout doesn't display subpages with the same layout.

- If the browser size is adjusted, the ellipsis in the menu doesn't disappear when it should and is displayed in the middle of the menu items or on some occasions on the left side of them.

- If there's only one language on the site, it's still displayed in the header for the visitor.

annes-aus reviewed (changes requested):
- Upon signup, new site has virtually no default content. Upside: user has a blank page with nothing limiting their imagination; downside: user might not get a good idea what each block is best suited for and may get distracted by the asymmetric blocks.
- Could use tooltips, especially considering it used numerous new UI elements not previously seen on Voog. For example: add new content block (plus sign), move block up/down, delete block.
- Move block up/down buttons not immediately noticeable; visually resemble expand/collapse buttons.
- Delete block button button small and not easily noticeable. On the other hand, clicking on it immediately deletes the block without a warning or without a possibility to undo.

PeeterPaal committed:
Add `postcss` + `autoprefixer` for compatibility

Add autoprefixer for improved browser compatibility, as for example
Safari does not support unprefixed backdrop-filter styles.

In addition, fix more bugs in Safari, e.g. menu background blur not
appearing in mobile view.

PeeterPaal committed:
Fix add block btn on Safari & Firefox

Use SVG for icon instead of plaintext

t-p-developer commented:
When this is settled and done, could we also add Percy's snapshot tests?

PeeterPaal commented:
> When this is settled and done, could we also add Percy's snapshot tests?

This can be discussed, adding these tests would require some extra steps to figure out the workflow.

PeeterPaal committed:
Blur header in mobile menu

Blur header background in mobile menu if menu is scrolled, in order
to not display menu items behind header.

PeeterPaal committed:
Lower header z-index

Make image resize bubbles appear in header content area.

PeeterPaal committed:
Improve main menu & langmenu behaviour

- Show language only if site has multiple languages
- Show item children in editmode only if item is translated
- Show submenus next to menuitems

PeeterPaal committed:
Improve handling menus on window resize

Make menu items invisble while window is being resized, so that menu
elements are not visible in wrong places while resizing.

In addition, fix dropdown menu ellipsis being visible after resizing
window left no items in it.

PeeterPaal committed:
Improve cart btn behaviour

- Do not show cart button if store is unavailable (in catalogue mode)
- Do not show cart button in mobile view if store is disabled or
unavailable

PeeterPaal committed:
Add confirmation dialog when deleting block

PeeterPaal committed:
Add titles to block edit buttons

PeeterPaal committed:
Simpligy front page layouts

Remove one content area from front page layout and adjust styles
accordingly

PeeterPaal committed:
Up header z-index

Fix add content area buttons coming through header

PeeterPaal committed:
Do not display block with no content in publicmode

PeeterPaal committed:
Improve auto-gen product page

- Add support for multi-assets by replacing product image with gallery
- Make product page at least 100% window height

PeeterPaal committed:
Fix menu items sometimes going into two rows

Wrapping flex items seems to happen between loading dom content and
fully loading window. Because of this, previously some edge cases were
missed when handling menu items. Wait for page to fully load before
handling menu items to account for this.

tanelj commented on components/footer.tpl:
```diff
@@ -0,0 +1,16 @@
+<footer class="footer">
```
Add edit mode titles.

```liquid
{%- assign footer_content_title_tooltip = "content_tooltip_all_pages_same_language" | lce -%}

{% xcontent name="footer-col-1" title=footer_content_title_tooltip %}
```

tanelj commented on components/header.tpl:
```diff
@@ -0,0 +1,38 @@
+    {% include "menu-main" %}
+
+    <div class="header-content content-formatted">
+      {% xcontent name="header" %}
```
Use edit mode hints (texts: https://www.voog.com/developers/markup/basics/i18n)
```suggestion
      {%- assign cross_site_title = "cross_site_title" | lce -%}
      {%- assign cross_site_title_tooltip = "content_tooltip_all_pages_same_language" | lce -%}
      {% xcontent name="header" title=cross_site_title tooltip=cross_site_title_tooltip %}
```

tanelj commented on components/bg-picker-scripts.tpl:
```diff
@@ -0,0 +1,40 @@
+  $(function() {
+    $('.bg-picker-area').each(function(index, pickerArea) {
+      var $picker = $(pickerArea).find('.bg-picker');
+      var pickerOpts = $picker.data();
```
Sanitize against missing `$picker`.

PeeterPaal committed:
Only hide menuitems on width change

tanelj commented on layouts/common_page.tpl:
```diff
@@ -0,0 +1,80 @@
+      {% endfor %}
+    </main>
+
+    {% if editmode %}
```
```suggestion
    {%- if editmode %}
```

tanelj commented on layouts/common_page.tpl:
```diff
@@ -0,0 +1,80 @@
+          {%- include "ico-cross" -%}
+        </button>
+      </div>
+    {% endif %}
```
```suggestion
    {% endif -%}
```

tanelj commented on layouts/common_page.tpl:
```diff
@@ -0,0 +1,80 @@
+    <main class="common-page-content" role="main" data-search-indexing-allowed="true">
+      {% assign allowed_layouts = template_settings.common_page_block_layouts | map: "key" %}
+
+      {% for block_data in body_blocks %}
```
```suggestion
      {%- for block_data in body_blocks %}
```

tanelj commented on layouts/common_page.tpl:
```diff
@@ -0,0 +1,80 @@
+          layout_name: layout_name,
+          layout_data: layout_data,
+          wrapper_class: "js-blocks-wrapper"
+        %}
```
```suggestion
        -%}
```

tanelj commented on components/block.tpl:
```diff
@@ -0,0 +1,126 @@
+  {%- endif -%}
+{%- endcapture -%}
+
+{%- if editmode -%}
+  {%- assign block_has_content = true -%}
+{%- else -%}
+  {%- capture block_contents_html -%}
+    {%- for index in (1..content_area_count) -%}
+      {%- assign block_area_name = "block-" | append: id | append: "-col-" | append: index -%}
+
+      {% content name=block_area_name %}
+    {%- endfor -%}
+  {%- endcapture -%}
+
+  {%- capture block_contents_size -%}
+    {{- block_contents_html | size | minus: 1 -}}
+  {%- endcapture -%}
+
+  {%- if block_contents_size contains "-" -%}
+    {%- assign block_has_content = false -%}
+  {%- else -%}
+    {%- assign block_has_content = true -%}
+  {%- endif -%}
+{%- endif -%}
+
+{%- if block_has_content %}
```
Can it be simpler?

```suggestion
{%- unless editmode -%}
  {%- capture block_contents_html -%}
    {%- for index in (1..content_area_count) -%}
      {%- assign block_area_name = "block-" | append: id | append: "-col-" | append: index -%}

      {% content name=block_area_name %}
    {%- endfor -%}
  {%- endcapture -%}
{%- endunless -%}

{%- if block_has_content or block_contents_html != blank %}
```

tanelj reviewed (changes requested):
Some feedback:

- [x] Search modadal z-index must be higher than site title area ("+" icon is shown)
<img width="1203" alt="Screenshot 2023-12-04 at 16 16 44" src="https://github.com/Voog/design-faro/assets/378755/932a7fa3-4fae-4f05-9a49-45989e43f38e">

- [x] Default table styles are missing
<img width="257" alt="Screenshot 2023-12-04 at 15 58 16" src="https://github.com/Voog/design-faro/assets/378755/75a8308b-8076-43ea-9495-a8b69e588010">


- [x] Mobile menu. Click on language name should also open a language menu
- [x] Mobile menu. Sub menu arrow should use pointer cursor and maybe optional hover tooltip
- [x] Mobile menu. Click on arrow doent't alway open sub menu.
- [x] Mobile menu. Clickable area of the sub menu icon (kebab) should be a bit bigger
- [x] Menu. Menu is not re-rendered when window height changes.
- [x] Content block layouts "Dominant right/left" values should be close in the menu:

<img width="205" alt="Screenshot 2023-12-04 at 16 57 04" src="https://github.com/Voog/design-faro/assets/378755/3538ab8a-661c-4e30-9190-0463f816e01e">

PeeterPaal committed:
Fix content area add btn visibility in searchmodal

PeeterPaal committed:
Add basic table styles

PeeterPaal committed:
Improve mobile menu handling

- Fix issues when clicking on menu children icon did not open submenu
- Show pointer when hovering on menu children icon

PeeterPaal committed:
Fix langmenu being visible in mobile submenu

PeeterPaal committed:
Increase kebabmenu hitbox on mobile

PeeterPaal committed:
Switch common page template order

PeeterPaal committed:
Add titles & tooltips to header & footer content

PeeterPaal committed:
Guard against missing bgpicker

PeeterPaal committed:
Improve stripping whitespaces

PeeterPaal committed:
Simplify figuring out if block has content

tanelj reviewed (changes requested):
Some feedback:

- [x] Mobile menu: Language menu should be centered
<img width="181" alt="Screenshot 2023-12-05 at 17 33 16" src="https://github.com/Voog/design-faro/assets/378755/957b3c08-aef7-4196-9673-44e3089848c9">

- [x] Form content are out of area in narrow screens

<img width="968" alt="Screenshot 2023-12-05 at 15 24 13" src="https://github.com/Voog/design-faro/assets/378755/f1f06861-51b2-40b4-b16c-d0318186d56b">

- [x] Search modal: Its hard to click lower part of close icon:
<img width="198" alt="Screenshot 2023-12-05 at 23 33 24" src="https://github.com/Voog/design-faro/assets/378755/3295ca3e-ee2b-4236-8b78-c1bec123a4f6">

PeeterPaal committed:
Fix form fields overflowing screen

PeeterPaal committed:
Improve mobile submenu arrow placement & langmenu

- Fix mobile submenu arrow being misaligned with menuitem
- Fix langmenu width in mobile menu

PeeterPaal committed:
Improve search behaviour

- Make sure search results are scrollable on lower height screens
- Make sure search modal close btn is not covered up by results
- Fix search icon alignment next to input

tanelj commented on layouts/front_page.tpl:
```diff
@@ -0,0 +1,87 @@
+
+        {%- if editmode -%}
+          <div class="change-layout-options">
+            <select class="js-change-layout-select" title="Change front page layout">
```
Maybe more general "Change content layout" is better (+ TODO: translation).

```suggestion
            <select class="js-change-layout-select" title="Change content layout">
```

tanelj commented on layouts/auto_generated_product.tpl:
```diff
@@ -0,0 +1,174 @@
+
+{%- unless footnote_content_size contains "-" -%}
+  {%- assign footnote_content_has_content = true -%}
+{%- endunless -%}
```
Because in edit mode there has alwasy content (wrapper html) then its better:
```suggestion
{%- capture footnote_content_html -%}
  {%- content bind=product name="footnote" -%}
{%- endcapture -%}
```

And later:
```
          {% if footnote_content_html != blank %}
            <div class="product-footnote">
              {{ footnote_content_html }}
            </div>
          {% endif %}
```

tanelj commented on layouts/auto_generated_product.tpl:
```diff
@@ -0,0 +1,174 @@
+
+{%- capture bottom_content_size -%}
+  {{- bottom_content_html | size | minus: 1 -}}
+{%- endcapture -%}
```
Better:
```suggestion
{%- capture bottom_content_html -%}
  {%- assign bottom_content_title = "additional_content" | lce -%}
  {%- assign bottom_content_title_tooltip = "content_tooltip_additional_information" | lce -%}
  {% content
    bind=product
    name="content"
    title=bottom_content_title
    title_tooltip=bottom_content_title_tooltip
  %}
{%- endcapture -%}
```

tanelj commented on layouts/auto_generated_product.tpl:
```diff
@@ -0,0 +1,174 @@
+  {%- assign bottom_content_has_content = true -%}
+{%- endunless -%}
+
+{%- capture product_social_html -%}
```
Same blank pattern can be used here.

tanelj commented on components/block-structure.tpl:
```diff
@@ -0,0 +1,106 @@
+
+{%- if bg_color == blank -%}
+  {%- assign bg_color = "none" -%}
+{%- endif -%}
```
It should be same as:
```suggestion
{%- assign bg_color = background.color | default: "none" -%}
```

tanelj commented on components/block.tpl:
```diff
@@ -0,0 +1,114 @@
+  {%- endcapture -%}
+{%- endunless -%}
+
+{%- if block_contents_html != blank or editmode %}
```
Maybe little bit better.
```suggestion
{%- unless editmode -%}
  {%- assign blocks_has_content = false -%}
  
  {%- for index in (1..content_area_count) -%}
    {%- assign block_area_name = "block-" | append: id | append: "-col-" | append: index -%}

    {%- capture _content_html -%}
      {% content name=block_area_name %}
    {%- endcapture -%}

    {%- if _content_html != blank -%}
      {%- assign blocks_has_content = true -%}
      {%- break -%}
    {%- endif -%}
  {%- endfor -%}
{%- endunless -%}

{%- if editmode or blocks_has_content %}
```

tanelj commented on components/footer.tpl:
```diff
@@ -0,0 +1,34 @@
+      %}
+    </div>
+  </div>
+</footer>
```
At the moment there is empty space in footer live mode when there actually has no content.

Maybe we should use here also 
```suggestion
{%- unless editmode -%}
  {%- assign footer_has_content = false -%}
  
  {%- for index in (1..4) -%}
    {%- assign content_name = "footer-col-" | append: index -%}

    {%- capture _content_html -%}
      {% content name=content_name %}
    {%- endcapture -%}

    {%- if _content_html != blank -%}
      {%- assign footer_has_content = true -%}
      {%- break -%}
    {%- endif -%}
  {%- endfor -%}
{%- endunless -%}

{%- if editmode or footer_has_content -%}
  <footer class="footer">
    <div class="footer-content">
      {%- assign footer_content_title = "cross_site_footer" | lce -%}
      {%- assign footer_content_title_tooltip = "content_tooltip_all_pages_same_language" | lce -%}
      {%- for index in (1..4) -%}
        {%- assign content_name = "footer-col-" | append: index -%}
        <div class="content-formatted">
          {% xcontent
            name=content_name
            title=footer_content_title
            title_tooltip=footer_content_title_tooltip
          %}
        </div>
      {%- endfor -%}
    </div>
  </footer>
{%- endif -%}
```

tanelj commented on components/template-meta.tpl:
```diff
@@ -0,0 +1,48 @@
+
+{%- if og_image -%}
+  {%- comment -%}"http:" and "https:" strings are removed and readded to ensure that older bg-picker images will have protocol.{%- endcomment -%}
+  {%- if og_image.url -%}<meta property="og:image" content="{{ og_image.url | replace_first: "http:", "" | replace_first: "https:", "" | prepend: "https:" }}">{%- endif -%}
```
It should be ok:
```suggestion
  {% if og_image.url %}<meta property="og:image" content="{{ og_image.url }}">{% endif %}
```

tanelj reviewed (changes requested):
Here are also some code related feedback.

PeeterPaal committed:
Improve rendering content on auto-gen pr. page

Simplify finding out whether area has content or not

PeeterPaal committed:
Simplify setting default bg color

PeeterPaal committed:
Improve rendering blocks & footer content

Do not render out footer if it has no content present. Simplify
finding if block or footer has content.

PeeterPaal committed:
Simplify setting og image meta property

PeeterPaal committed:
Add hover action to search result

PeeterPaal committed:
Use system translations

PeeterPaal committed:
Always open mobile submenu in editmode

As subpage add button is only visible in submenu, it needs to be always
openable in editmode.

PeeterPaal committed:
Align front page split columns to top

vkjgr commented:
Hi!

Desktop

- [ ] As the block has a Background color picker icon in the top left corner and a Trashbin in the top right corner, I assume that the Trashbin will delete the image. There is a TITLE, but it has a browser delay. Some possible solutions: Can we add a tooltip on hover or write a better alert text? Something like “Soovid kustutada blokki. Oled sa kindel?”
- [ ] When adding text blocks, the initial text color is white. After adding another one, it turns black. It should not change. I got it. When adding a background image, the text color turns white.
- [ ] How can I add an image as the Site Title without adding a Gallery? Like in the theme demo.
- [ ] Is it intentional that a single row is half the width, not the page width?
- [ ] There are only 2 menu levels accessible.

Mobile

- [ ] In Figma, the header-content is aligned left, but in construction it is centered. Both are nice, I just noticed the difference. Is it because of the second menu back button?
- [ ] The second level menu should come from the right :)
- [ ] I would like to see the cart always in the menu. At least the number of items in it. See note:
- [ ]  Let’s speed up the transitions from 1s to 300ms

Best practices for shopping cart design
1. Place the shopping cart icon in the top right corner of the page
https://webflow.com/blog/ecommerce-cart-design

PeeterPaal commented:
Desktop:

> There is a TITLE, but it has a browser delay. Some possible solutions: Can we add a tooltip on hover or write a better alert text?

We would like to keep the alert text as general as possible, as these are system-wide translations. I will explore some options to see whether adding a tooltip that appears faster on hover is possible.

> When adding text blocks, the initial text color is white. After adding another one, it turns black. It should not change. 

Text colors automatically adjust based on content background, perhaps that was the case here?

> How can I add an image as the Site Title without adding a Gallery? Like in the theme demo.

By adding a Text type content area and drag-dropping an image into that :)

> Is it intentional that a single row is half the width, not the page width?

In design the "hero" content in Front page is not very wide, which is why it is only half-width. 

> There are only 2 menu levels accessible.

Currently it is planned to only support maximum of 2 levels in menus. Most standard designs follow the same pattern, but if need arises, then adding lower levels can be considered.

Mobile:

> In Figma, the header-content is aligned left, but in construction it is centered. Both are nice, I just noticed the difference. Is it because of the second menu back button?

Yes, I figured that the header content should perhaps always be visible, even with the back arrow present.

> The second level menu should come from the right :)

Will change that.

> I would like to see the cart always in the menu.

I'll take a look if it can be fixed to the bottom of the screen, but as it is quite large, it can take up a lot of screen space. In Figma Proto, the Cart button was not fixed to the bottom, which is why it is currently implemented like that. 

I'll can also play around with displaying the current number of products in cart.

> Let’s speed up the transitions from 1s to 300ms

Will change that too :)

PeeterPaal committed:
Add better tooltips to editmode icons

Add tooltips that appear on hover to editmode icons to make their
functionality more clear.

PeeterPaal committed:
Make mobile menu animations faster

Make mobile menu animations run faster & make mobile submenu open
from right side.

PeeterPaal committed:
Fixate cart btn to the end of the menu

PeeterPaal committed:
Improve auto-gen pr. page word wrap

PeeterPaal committed:
Add `theme-color` meta tag

Include `theme-color` meta tag to help browsers use background colour
for browser UI elements.

Reduce mobile submenu background opacity for better experience on
mobile Safari.

PeeterPaal committed:
Fix darkmode colors for cart button

PeeterPaal committed:
Fix search & cart btn paddings on desktop

PeeterPaal committed:
Show cart items count on menu burger in mobile

tanelj commented on components/template-settings-json.tpl:
```diff
@@ -71,13 +71,13 @@
-    "split_60_40": "Dominant left",
-    "split_staggered": "Staggered",
-    "split_50_50_full": "Full width"
+    "split_even": "{{ "split" | lce }}",
+    "column": "{{ "column" | lce }}",
+    "split_dominant_right": "{{ "dominant_right" | lce }}",
+    "split_40_60": "{{ "dominant_right" | lce }}",
+    "split_60_40": "{{ "dominant_left" | lce }}",
+    "split_staggered": "{{ "staggered" | lce }}",
+    "split_50_50_full": "{{ "full_height" | lce }}"
```
Maybe iits more safe code, since translations can contain "` or `'`
```suggestion
    "split_even": {{ "split" | lce | json }},
    "column": {{ "column" | lce | json }},
    "split_dominant_right": {{ "dominant_right" | lce | json }},
    "split_40_60": {{ "dominant_right" | lce | json }},
    "split_60_40": {{ "dominant_left" | lce | json }},
    "split_staggered": {{ "staggered" | lce | json }},
    "split_50_50_full": {{ "full_height" | lce | json }}
```

tanelj commented on sources/stylesheets/_general.scss:
```diff
@@ -47,11 +50,33 @@ main {
+    opacity: 0;
+    transition: opacity .3s ease-in-out;
+    padding: 8px;
+    width: fit-content;
```
In Firefox its wrapped:

![Screenshot 2023-12-11 at 18 27 52](https://github.com/Voog/design-faro/assets/378755/e968ad48-7433-4055-8550-664fd8450f50)
![Screenshot 2023-12-11 at 18 27 45](https://github.com/Voog/design-faro/assets/378755/e68e2819-11ba-49b2-9a43-c84e767b46eb)

tanelj reviewed (changes requested):
Nice work but there has som minor issues.

Also it seems that in front page layout footer "flying" when there has few content on main area.

![Screenshot 2023-12-11 at 18 31 03](https://github.com/Voog/design-faro/assets/378755/65553a8a-e7ea-463f-9f07-ac81d85dde31)

PeeterPaal committed:
Render translations with `json` filter

PeeterPaal committed:
Fix editmode tooltips word wrap

PeeterPaal committed:
Fix word wrap in dropdown menus

PeeterPaal committed:
Improve page behaviour with little to no content

Show page background color only until footer, after which footer bg
color is displayed until the end of the page.

PeeterPaal committed:
Do not render out empty product list content

PeeterPaal committed:
Animate common page blocks

Use IntersectionObserver to figure out whether block is in viewport
and run animations if block enters viewport for the first time.

PeeterPaal committed:
Simplify common block animations

- Use fixed values for animations to ensure better consistency
- Simplify CSS rules for animations as a result

PeeterPaal committed:
Use center alignment for front page split sections

PeeterPaal committed:
Lower front page hero content min-height on mobile

PeeterPaal committed:
Add gap between fr. page split section in mobile

tanelj reviewed (approved):
Nice work!

PeeterPaal committed:
Change default styles

- Lower H1 default font size
- Do not underline links in footer by default

tanelj reviewed (approved):
Nice!
